File: apps/edge/src/redis.ts
import IORedis from 'ioredis';
import { config } from './config.js';

let _client: IORedis | null = null;

export function getRedis(): IORedis {
  if (_client) return _client;
  _client = new IORedis(config.REDIS_URL, {
    lazyConnect: false,
    maxRetriesPerRequest: 3,
    enableReadyCheck: true,
  });
  return _client;
}

export async function isIdempotentlyNew(key: string, ttlSeconds = 86_400): Promise<boolean> {
  const r = getRedis();
  const set = await r.set(key, '1', 'EX', ttlSeconds, 'NX');
  return set === 'OK';
}

export function tenantKey(tenantId: string, ...parts: string[]): string {
  return ['tenant', tenantId, ...parts].join(':');
}

export function globalKey(...parts: string[]): string {
  return ['global', ...parts].join(':');
}

File: packages/python-shared/veda_shared/infra/redis.py
"""Async Redis client (singleton)."""

from __future__ import annotations

import redis.asyncio as redis

from veda_shared.settings import get_settings

_client: redis.Redis | None = None


def get_redis() -> redis.Redis:
    global _client
    if _client is None:
        _client = redis.from_url(get_settings().redis_url, decode_responses=True)
    return _client


def tenant_key(tenant_id: str, *parts: str) -> str:
    return ":".join(["tenant", tenant_id, *parts])


def global_key(*parts: str) -> str:
    return ":".join(["global", *parts])
